enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,577 @@
+and `operatorsPackagesAndChannels` fields are nested under the `PreCachingConfig.spec.overrides` object, it
+is important to note that they will still be shown as a flat json object under `status.precaching` in the CGU CR.
+
+```json
+{
+...
+  "status": {
+    "conditions": [...],
+    "precaching": {
+        "clusters": ["sno1", "sno2"],
+        "spec": {
+          "platformImage": "quay.io/openshift-release-dev/ocp-release@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef"},
+        "status": {
+          "sno1": "Active",
+          "sno2": "Starting"}
+    }
+  }
+}
+```
+
+The pre-caching CR needs to be referenced in the CGU. There are three potential ways in which this can be
+done. Each option carries its own merrit and challenges.
+
+**Option 1:**
+This option is the simplest solution. It entails creating a new top-level field under the CGU `spec`
+object to reference the pre-caching config CR. In this approach, the `PreCachingConfig` CR is referenced
+in the CGU under the field name `preCachingConfigRef`. The `preCaching` field is still used to enable the
+pre-caching job.
+
+This approach is shown below.
+
+```yaml
+---
+apiVersion: ran.openshift.io/v1alpha1
+kind: ClusterGroupUpgrade
+metadata:
+  name: cgu
+spec:
+  preCaching: true
+  preCachingConfigRef: foobar
+...
+```
+
+**Option 2:**
+This option entails creating a new top-level pre-caching field in the CGU named `preCache`. This field
+comprises of the `enable` and the `configRef` fields. The former is used to enable/disable pre-caching
+whereas the latter is used to specify the reference to the pre-caching configuration CR.
+
+The intention behind this approach is to contain all pre-caching fields under one object and to
+deprecate the current `preCaching` field. Note that if users continue to use the original `preCaching`
+field - that is fine. However, the `precache.enable` field will take precedence (or cause failure if
+both are set and not the same).
+
+This approach is shown below.
+
+```yaml
+---
+apiVersion: ran.openshift.io/v1alpha1
+kind: ClusterGroupUpgrade
+metadata:
+  name: cgu
+spec:
+  preCaching: true # to be deprecated in future releases
+  preCache:
+    enable: true
+    configRef: foobar
+...
+```
+
+**Option 3:**
+This option offers an idealistic solution by avoiding multiple pre-caching fields by modifying the
+top-level `preCaching` CGU spec field from boolean type to map type. This approach requires transitioning
+to a new `apiVersion` for the CGU CR (`apiVersion: ran.openshift.io/v2alpha1`).
+
+A major drawback to this approach is that we need to ensure backwards compatibility, in other words, we
+would need to support both `/v1alpha1` and `/v2alpha1` versions of the API simultaneously.
+
+This approach is shown below.
+
+```yaml
+---
+apiVersion: ran.openshift.io/v2alpha1
+kind: ClusterGroupUpgrade
+metadata:
+  name: cgu
+spec:
+  preCaching:
+    enable: true
+    configRef: foobar
+...
+```
+
```

Comment:
Although I prefer option 3, I'm not convinced it is worth a full up-rev of the version for the incremental "cleanness" over option 2. 
My order of choice would be option 2, 3, and then a much further back 1.

Reply:
I would think option 1 is more palatable because it is backward compatible (contrary to 2 and 3) and less confusing to users (contrary to the 2). That's right, it's not the cleanest one. But API change is a move that should be planned for, due to the overhead it bears on all levels (us, QE, docs, users). Are there other API changes we consider? Maybe we should bundle them all in a change unrelated to this enhancement?

Reply:
Option 1,2 do an incremental change in the API, and 1 is the most backward compatible and cleaner, just referencing the precaching CR. Option 3 needs API redesign, not backward compatible and can be considered a good approach when a major API change is planned. I would vote in order of preference ->1 -> 2.

Reply:
Thank you all for your feedback.  The current standings on how to reference the new CR appears to be as follows: Option 1: (Vitaly, Jun, Sabbir), Option 2 (Ian M.), Option 3 (none). There is a majority consensus on Option 1. I will update the PR to reflect Option 1 as the selected approach.

Reply:
FWIW, I also feel that option 1 _might_ be easier to adopt by users, and cleaner as well. Additionally, by just adding the field `precachingConfigRef`, don't see it as an over-abuse of the word -precaching- in the API spec either, imho.

Reply:
My reasoning for putting option 1 at the end of the list was to avoid continued expansion of the top-level elements of the configuration. I agree with the thought that having a precaching CR now mitigates future expansion of the top level, though I don't think it eliminates it (eg future precaching "continue on failure" setting, etc). Option 2 sets us on a path where we have the proper structure to handle that in the future. 

That said, I think the risk of future expansion is low and if it does occur we still have the ability to move to option 2 at that point. I'm ok if we go with option 1.

Reply:
Initially I envisioned a flat precachingSpec CR which is more decoupled from TALM and CGU behavior. I would have agreed with your concern on something like "continue on failure" if we had chosen that path. Now that we've brought the override concept in, this CR has become more of a CGU sub type. I don't see anything wrong with adding the failure action to this precachingConfig CR.

Reply:
I agree with Jun, the continue-on-failure option, and any other pre-caching specific configuration can, and should, be defined in the PreCachingConfig CR.